Improve the various groups libraries

fingolfin opened the pull request:
<s>This is work in progress. Still various TODOs and FIXMEs need to be resolved.
But I want this out there so it doesn't get neglected for another 6 months :-(</s>

Should be ready now. More should be done, and there are some TODOs being added, but I think it's still overall an improvement.

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -358,13 +358,12 @@ end
 Return (`true`,`f`) if `G` and `H` are isomorphic groups, where `f` is a group
-isomorphism. Otherwise, return (`false`,`f`), where `f` is the trivial
-homomorphism.
+isomorphism. Otherwise, return (`false`,`nothing`).
```
this doesn't really belong in this PR, but: I hope this minor type instability is OK for people, but I found it utterly confusing to get a "trivial homomorphism" in case of a failure.

But really, I think `isisomorphic` should return just a single boolean by default, and return the iso only when requested, resp. there should be a separate function that is called when one wants the iso...

fingolfin replied:
See PR #1246

fingolfin committed:
Improve the various groups libraries

ThomasBreuer commented on docs/src/Groups/grouplib.md:
```diff
@@ -17,19 +17,39 @@ Pages = ["grouplib.md"]
+
+The arrangement and the names of the groups of degree up to 15 is the same as given in
+[CHM98](@cite). With the exception of the symmetric and alternating group (which are represented
+as `SymmetricGroup` and `AlternatingGroup`) the generators for these groups also conform to this paper with
```
`SymmetricGroup` and `AlternatingGroup` are the GAP function names, should we write `symmetric_group` and `alternating_group` here?

ThomasBreuer commented on src/Groups/libraries/primitivegroups.jl:
```diff
@@ -1,27 +1,55 @@
-Return the `i`-th group in the catalogue of primitive groups of degree `n`
-in GAP's Primitive Groups Library. The output is a group of type `PermGroup`.
+Return the `i`-th group in the catalogue of primitive groups over the set
+{`1`,...,`deg`} in the GAP Small Groups Library. The output is a group of type
```
in GAP's library of primitive groups

ThomasBreuer commented on src/Groups/libraries/primitivegroups.jl:
```diff
@@ -31,45 +59,28 @@ end
-and the function returns all the groups `G` satisfying the conditions
-`func1(G) == arg1`, `func2(G) == arg2`, etc.
-An argument can be omitted if it corresponds to the boolean value `true`.
+satisfying the conditions describe by the arguments. These conditions
```
described

ThomasBreuer commented on src/Groups/libraries/smallgroups.jl:
```diff
@@ -1,89 +1,198 @@
-  return (r[1], r[2])
+   isfinite(G) || error("group is not finite")
+   res = GAP.Globals.IdGroup(G.X)
+   res !== GAP.Globals.fail || error("identification is not available for groups of order $(order(G))")
```
Perhaps check directly with `has_small_group_ids`?

fingolfin replied:
Aye. And perhaps `has_small_group_ids` should get a better name... Maybe `has_small_group_identification` or `small_group_identification_available` or so? Hm

fingolfin replied:
Hrm, I just realized I added `transitive_groups_available` but perhaps it would make more sense / fit in better with the rest to call that `has_transitive_groups`

ThomasBreuer commented on src/Groups/libraries/smallgroups.jl:
```diff
@@ -1,89 +1,198 @@
+   isfinite(G) || error("group is not finite")
+   res = GAP.Globals.IdGroup(G.X)
+   res !== GAP.Globals.fail || error("identification is not available for groups of order $(order(G))")
+   return Tuple{Int,Int}(res)
```
Is it perhaps inconsistent if the result of `number_small_groups` is an `fmpz` but the second entry of the id is an `Int`?

fingolfin replied:
Indeed! For that matter, one could argue that the group order should also be an fmpz, after all, one can do `NumberSmallGroups(NextPrimeInt(2^64));` in GAP just fine, so why not in Oscar...

ThomasBreuer commented on src/Groups/libraries/smallgroups.jl:
```diff
@@ -1,89 +1,198 @@
+Note that at least one of the conditions must impose a limit on the group
+order, otherwise an error is raised.
+
+The type of the returned groups is `PcGroup` if the group is solvable, `PermGroup` otherwise.
```
Eventually (not in this pull request) we can support a type as the first argument of `all_small_groups`, with the same meaning as in `small_group`.

ThomasBreuer commented on src/Groups/libraries/transitivegroups.jl:
```diff
@@ -1,17 +1,33 @@
+"""
+function number_transitive_groups(deg::Int)
+   res = GAP.Globals.NrTransitiveGroups(deg)
+   res isa Int && return res
```
`Int` or `fmpz`?

ThomasBreuer commented on src/Groups/libraries/transitivegroups.jl:
```diff
@@ -31,11 +48,10 @@ end
-If `G` moves more than 31 points then `-1` is returned.
-Otherwise, if `G` is not transitive on its moved points then `0` is returned,
+If `G` is not transitive, or if the transitive groups of degree `d`
+are not available, an exception is raised.
```
an exception is thrown

ThomasBreuer commented on src/Groups/libraries/transitivegroups.jl:
```diff
@@ -61,38 +77,44 @@ julia> order(transitive_group(4, m)) == order(S)
-`arg1`, `func2`, `arg2`, ... ], and the function returns all the groups `G`
-satisfying the conditions `func1`(`G`) = `arg1`, `func2`(`G`) = `arg2`, etc.
-An argument can be omitted if it corresponds to the boolean value `true`.
+satisfying the conditions describe by the arguments. These conditions
```
described

ThomasBreuer commented on src/Groups/libraries/smallgroups.jl:
```diff
@@ -1,89 +1,198 @@
-`arg2`, etc. An argument can be omitted if it corresponds to the boolean value
-`true`.
+Return the list of all groups (up to isomorphism)
+satisfying the conditions describe by the arguments. These conditions
```
described

ThomasBreuer reviewed (approved):
Looks good, thanks. I have added a few minor remarks.
(Just now I noticed that GAP does not provide `AllPerfectGroups` etc.)

fingolfin committed:
Further improve groups libraries

fingolfin commented on docs/src/Groups/grouplib.md:
````diff
@@ -59,19 +84,29 @@ not have a faithful permutation representation of small degree.
 
 ```@docs
-orders_perfect_groups
+has_number_perfect_groups
+has_perfect_group_identification
````
These two still need to be implemented

fingolfin replied:
done

fingolfin committed:
Drop support for old all_*_groups syntax

E.g. all_transitive_groups(order, 6) no longer is accepted, use all_transitive_groups(order => 6) instead

ThomasBreuer reviewed (approved):
Very nice.
